### PR TITLE
Infeasibility analyzer : improve unit tests 

### PR DESCRIPTION
**Pull request by guilpier-code**
Mainly add new unit tests for slack analysis.
For details, some **gitHub** comments were left.

### CODE REVIEW

File: src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ void ConstraintSlackAnalysis::run(MPSolver* problem)
     addSlackVariablesToConstraints(problem);
     if (slackVariables_.empty())
     {
-        logs.error() << title() << " : no constraints have been selected";
+        logs.warning() << title() << " : no constraints have been selected";
```
Comment (guilpier-code) [inline src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:53]: If no constraint was selected for analysis, do we have to raise an error ? Or a warning's enough ?

File: src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp [hunk 2]
```diff
@@ -63,10 +63,12 @@ void ConstraintSlackAnalysis::run(MPSolver* problem)
         return;
     }
 
-    hasDetectedInfeasibilityCause_ = true;
-
     sortSlackVariablesByValue();
     trimSlackVariables();
+    if (hasDetectedInfeasibilityCause_ = anySlackVariableNonZero(); !hasDetectedInfeasibilityCause_)
```
Comment (guilpier-code) [inline src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:68]: New behavior : print a message when no constraint violation is detected

File: src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp [hunk 3]
```diff
@@ -130,11 +132,24 @@ void ConstraintSlackAnalysis::trimSlackVariables()
...
+}
+
 void ConstraintSlackAnalysis::printReport() const
 {
     InfeasibleProblemReport report(slackVariables_);
-    report.logSuspiciousConstraints();
-    report.logInfeasibilityCauses();
+    report.storeSuspiciousConstraints();
+    report.storeInfeasibilityCauses();
+    std::ranges::for_each(report.getLogs(), [](auto& line) { logs.notice() << line; });
```
Comment (guilpier-code) [inline src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:152]: For test purpose, the report object now produces a vector of logs (can be retrieved with method **getLogs()**).
And indeed, logging is not supposed to be done by a report builder (here an instance of **InfeasibleProblemReport**).

File: src/solver/infeasible-problem-analysis/watched-constraints.cpp [hunk 4]
```diff
@@ -36,7 +36,7 @@ std::string StringBetweenAngleBrackets(const std::string& constraintName)
 
 std::string timeStep(std::vector<std::string> splitName)
 {
-    return StringBetweenAngleBrackets(splitName.at(splitName.size() - 2));
+    return StringBetweenAngleBrackets(splitName.rbegin()[1]);
```
Comment (guilpier-code) [inline src/solver/infeasible-problem-analysis/watched-constraints.cpp:39]: We could even call `splitName.end()[-2]` instead.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 5]
```diff
@@ -203,16 +238,109 @@ BOOST_AUTO_TEST_CASE(analysis_should_ignore_ill_named_constraint)
...
+    std::unique_ptr<MPSolver> feasibleProblem = createFeasibleProblem("BC-name::hourly::hour<36>");
     BOOST_CHECK(feasibleProblem->Solve() == MPSolver::OPTIMAL);
 
     ConstraintSlackAnalysis analysis;
     analysis.run(feasibleProblem.get());
-    BOOST_CHECK(analysis.hasDetectedInfeasibilityCause()); // Would expect false here instead?
+    BOOST_CHECK(!analysis.hasDetectedInfeasibilityCause());
+}
+
+BOOST_AUTO_TEST_CASE(analysis_slack_variables_are_ordered)
```
Comment (guilpier-code) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:227]: Here some new unit tests were added : 
- Unit tests for slack analysis class (the result of this class [at end of method **run()**]  is a collection of slack variable, ordered by slack value).
  - check that slack variables collection is ordered in expected order
  - check that slack variables collection is limited to 10 elements
- Unit tests for the report builder
  - all violated incoming constraints are reported
  - infeasibility causes are unique and sorted by slack value

File: src/solver/infeasible-problem-analysis/report.cpp [hunk 6]
```diff
@@ -24,8 +24,6 @@
 #include <regex>
 #include <typeindex>
 
-#include <antares/logs/logs.h>
-
 namespace Antares::Optimization
 {
 InfeasibleProblemReport::InfeasibleProblemReport(
```
Comment (guilpier-code) [inline src/solver/infeasible-problem-analysis/report.cpp:29]: class **InfeasibleProblemReport** : 
Actually, what it does is : build a report for slack variable analysis.
So a better name could be : **SlackAnalysisReportBuilder**.
The associated file name could change accordingly.

File: src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp (hunk 1 above)
Comment (flomnes) [reply src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:53]: Yes, good enough

File: src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/constraint-slack-analysis.h [hunk 7]
```diff
@@ -52,15 +52,19 @@ class ConstraintSlackAnalysis: public UnfeasibilityAnalysis
         return "Slack variables analysis";
     }
 
+    std::vector<const operations_research::MPVariable*>& largestSlackVariables();
```
Comment (flomnes) [inline src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/constraint-slack-analysis.h]:
```suggestion
    const std::vector<const operations_research::MPVariable*>& largestSlackVariables();
```

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 8]
```diff
@@ -21,13 +21,16 @@
 #define WIN32_LEAN_AND_MEAN
 #define BOOST_TEST_MODULE unfeasible_problem_analyzer
 
+#include <cstdio>
```
Comment (flomnes) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]:
```suggestion
```

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 9]
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
...
+    constraint->SetBounds(ConstLowBnd, problem->infinity());
+}
+
+std::unique_ptr<MPSolver> createProblemWith_n_violatedConstraints(const int n)
+{
+    std::unique_ptr<MPSolver> problem(MPSolver::CreateSolver("GLOP"));
+    for (auto i: std::ranges::iota_view(1, n + 1)) // From 1 to n included
+    {
+        char name[32];
+        std::sprintf(name, "BC-name-%d::hourly::hour<%d>", i, 5 * i);
```
Comment (flomnes) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Use `std::ostringstream` or `std::to_string` ? `sprintf` is C

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 10]
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
...
+                                  const std::string& name,
+                                  const double& varLowBnd,
+                                  const double& varUpBnd,
+                                  const double& ConstLowBnd)
+{
+    std::string varName = "slack-for-" + name;
+    auto* var = problem->MakeNumVar(varLowBnd, varUpBnd, varName);
+    auto* constraint = problem->MakeRowConstraint(name);
+    constraint->SetCoefficient(var, 1);
+    constraint->SetBounds(ConstLowBnd, problem->infinity());
```
Comment (flomnes) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]:
```suggestion
    auto* constraint = problem->MakeRowConstraint(ConstLowBnd, problem->infinity(), name);
    constraint->SetCoefficient(var, 1);
```

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 11]
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
...
+std::unique_ptr<MPSolver> createProblem(const std::string& constraintName)
+{
+    std::unique_ptr<MPSolver> problem(MPSolver::CreateSolver("GLOP"));
+    const double infinity = problem->infinity();
+    problem->MakeNumVar(1, infinity, "var1");
+    problem->MakeNumVar(-infinity, -1, "var2");
+    auto constraint = problem->MakeRowConstraint(constraintName);
+    constraint->SetBounds(0, infinity);
+    return problem;
+}
```
Comment (flomnes) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]:
```suggestion
std::tuple<std::unique_ptr<MPSolver>, MPVariable*, MPVariable*> createProblem(const std::string& constraintName)
{
    std::unique_ptr<MPSolver> problem(MPSolver::CreateSolver("GLOP"));
    const double infinity = problem->infinity();
    MPVariable* var1 = problem->MakeNumVar(1, infinity, "var1");
    MPVariable* var2 = problem->MakeNumVar(-infinity, -1, "var2");
    auto constraint = problem->MakeRowConstraint(constraintName);
    constraint->SetBounds(0, infinity);
    return {problem, var1, var2};
}
```
So you don't need `LookupVariableOrNull` afterwards.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 12]
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
...
+                                  const double& ConstLowBnd)
+{
+    std::string varName = "slack-for-" + name;
+    auto* var = problem->MakeNumVar(varLowBnd, varUpBnd, varName);
+    auto* constraint = problem->MakeRowConstraint(name);
+    constraint->SetCoefficient(var, 1);
+    constraint->SetBounds(ConstLowBnd, problem->infinity());
+}
+
+std::unique_ptr<MPSolver> createProblemWith_n_violatedConstraints(const int n)
```
Comment (flomnes) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:118]:
```suggestion
std::unique_ptr<MPSolver> createProblemWith_n_MutuallyIncompatibleConstraints(const int n)
```

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp [hunk 13]
```diff
@@ -83,7 +87,73 @@ class AnalysisMock: public UnfeasibilityAnalysis
...
+    constraint->SetCoefficient(var1, -1);
+    constraint->SetCoefficient(var2, 1);
+    return problem;
+}
+
+void addOneVarConstraintToProblem(MPSolver* problem,
+                                  const std::string& name,
+                                  const double& varLowBnd,
+                                  const double& varUpBnd,
+                                  const double& ConstLowBnd)
```
Comment (flomnes) [inline src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Don't use const references for `double`

**Review by flomnes — COMMENTED** [review]
Please pass POD types by non-const value.

Consider using fixtures for tests.

**Comment by guilpier-code** [comment]
> Consider using fixtures for tests.

I simplified unit tests a bit with last commit, and in the actual state, I can't see any fixture to define : 
For me, fixture : 
- are only related to setting test preparation like building test input data
- are used to avoid duplication

I cant' see no such things in the tests we have here.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 9 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Yes. Originally I used **std::format** on **Win**, unfortunately not available on Linux.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 11 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: I replaced this code by some other code, simpler, and function **createProblem(...)** is now gone.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 10 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2249#discussion_r1681284477), it applies here as well.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 12 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp:118]: What do you mean ? 
These constraints are unfeasible independently from one another : take only one of them, it's infeasible on its own.

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 9 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Done by using **std::stringstream**

File: src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/constraint-slack-analysis.h (hunk 7 above)
Comment (guilpier-code) [reply src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/constraint-slack-analysis.h]: Done

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 13 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Corrected

File: src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp (hunk 8 above)
Comment (guilpier-code) [reply src/tests/src/solver/infeasible-problem-analysis/test-unfeasible-problem-analyzer.cpp]: Done

**Review by flomnes — APPROVED** [review]

**Review by flomnes — APPROVED** [review]

**Pull request merged**